core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java

```diff
@@ -19,23 +19,665 @@
+
+        final DateTime t2 = new DateTime(2016, 12, 25, 7, 59, 42, 213, DateTimeZone.UTC);
+        assertResult(() -> builder().startObject().field("t2", t2).endObject(),
+                equalToJSon("{'t2':'2016-12-25T07:59:42.213Z'}"));
+        assertResult(() -> builder().startObject().field("t2").value(t2).endObject(),
+                equalToJSon("{'t2':'2016-12-25T07:59:42.213Z'}"));
+
+        final DateTimeFormatter formatter = randomFrom(ISODateTimeFormat.basicDate(), ISODateTimeFormat.dateTimeNoMillis());
+        final DateTime t3 = DateTime.now();
+        assertResult(() -> builder().startObject().field("t3", t3, formatter).endObject(),
+                equalToJSon("{'t3':'" + formatter.print(t3) + "'}"));
+        assertResult(() -> builder().startObject().field("t3").value(t3, formatter).endObject(),
+                equalToJSon("{'t3':'" + formatter.print(t3) + "'}"));
+
+        expectThrows(IllegalArgumentException.class, () -> builder().startObject().field("t3", t3, null).endObject());
+        expectThrows(IllegalArgumentException.class, () -> builder().startObject().field("t3").value(t3, null).endObject());
+    }
+
+    public void testDate() throws Exception {
+        assertResult(() -> builder().startObject().field("date", (Date) null).endObject(),
+                equalToJSon("{'date':null}"));
+        assertResult(() -> builder().startObject().field("date").value((Date) null).endObject(),
+                equalToJSon("{'date':null}"));
+
+        final Date d1 = new DateTime(2016, 1, 1, 0, 0, DateTimeZone.UTC).toDate();
+        assertResult(() -> builder().startObject().field("d1", d1).endObject(),
+                equalToJSon("{'d1':'2016-01-01T00:00:00.000Z'}"));
+        assertResult(() -> builder().startObject().field("d1").value(d1).endObject(),
+                equalToJSon("{'d1':'2016-01-01T00:00:00.000Z'}"));
+
+        final Date d2 = new DateTime(2016, 12, 25, 7, 59, 42, 213, DateTimeZone.UTC).toDate();
+        assertResult(() -> builder().startObject().field("d2", d2).endObject(),
+                equalToJSon("{'d2':'2016-12-25T07:59:42.213Z'}"));
+        assertResult(() -> builder().startObject().field("d2").value(d2).endObject(),
+                equalToJSon("{'d2':'2016-12-25T07:59:42.213Z'}"));
+
+        final DateTimeFormatter formatter = randomFrom(ISODateTimeFormat.basicDate(), ISODateTimeFormat.dateTimeNoMillis());
+        final Date d3 = DateTime.now().toDate();
+        assertResult(() -> builder().startObject().field("d3", d3, formatter).endObject(),
+                equalToJSon("{'d3':'" + formatter.print(d3.getTime()) + "'}"));
+        assertResult(() -> builder().startObject().field("d3").value(d3, formatter).endObject(),
+                equalToJSon("{'d3':'" + formatter.print(d3.getTime()) + "'}"));
+
+        expectThrows(IllegalArgumentException.class, () -> builder().startObject().field("d3", d3, null).endObject());
```

Comment:
Can you assert something about the message? As it stands someone could break something to throw a **different** `IllegalArgumentException` and we'd not notice.